src/ompl/base/samplers/informed/src/PathLengthDirectInfSampler.cpp

```diff
@@ -0,0 +1,357 @@
+*     with the distribution.
+*   * Neither the name of the University of Toronto nor the names of its
+*     contributors may be used to endorse or promote products derived
+*     from this software without specific prior written permission.
+*
+*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+*  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+*  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+*  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+*  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+*  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+*  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+*  POSSIBILITY OF SUCH DAMAGE.
+*********************************************************************/
+
+/* Authors: Jonathan Gammell */
+
+#include "ompl/base/samplers/informed/PathLengthDirectInfSampler.h"
+#include "ompl/util/Exception.h"
+#include "ompl/base/OptimizationObjective.h"
+#include "ompl/base/goals/GoalState.h"
+#include "ompl/base/StateSpace.h"
+#include "ompl/base/spaces/RealVectorStateSpace.h"
+
+//For strncmp
+#include <string.h>
+//For boost::make_shared
+#include "boost/make_shared.hpp"
+
+namespace ompl
+{
+    namespace base
+    {
+        //The direct ellipsoid sampling class for path-length:
+        PathLengthDirectInfSampler::PathLengthDirectInfSampler(const StateSpace* space, const ProblemDefinitionPtr probDefn, const Cost* bestCost)
+          : InformedStateSampler(space, probDefn, bestCost),
+            informedIdx_(0u),
+            uninformedIdx_(0u)
+        {
+            //Variables
+            //The foci of the ellipse as State* s
+            State* startFocusState;
+            State* goalFocusState;
+            //The foci of the ellipse as std::vectors
+            std::vector<double> startFocusVector;
+            std::vector<double> goalFocusVector;
+
+            //Sanity check the problem.
+            if (probDefn_->getStartStateCount() != 1u)
+            {
+                throw Exception("The direct path-length informed sampler currently only supports 1 start state.");
+            }
+
+            if (probDefn_->getGoal()->hasType(GOAL_STATE) == false)
+            {
+                throw Exception("The direct path-length informed sampler currently only supports goals that can be cast to goal states.");
+            }
+
+            //Check that the provided statespace is compatible and extract the necessary indices.
+            //The statespace must either be R^n or SE(2) or SE(3)
+            if (StateSampler::space_->isCompound() == false)
```

Comment:
Ideally, you could use space_->type_ to check if a space is the right type, but derived classes may not change the type_. Maybe typeid(T) from the <typeinfo> header is a more efficient way to check if the space is of the right type?